File: spo-kernel/crates/spo-engine/src/carrier.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Phase Orchestrator — SSGF geometry carrier W(t)

//! Softplus decode z → W and finite-difference gradient for the SSGF
//! outer cycle.

/// Decode latent vector z → coupling matrix W via softplus(A·z).
///
/// W_ij = log(1 + exp(A·z))_ij, diagonal zeroed.
///
/// # Arguments
/// * `z` – latent vector, length `z_dim`
/// * `a` – projection matrix, row-major (n*n × z_dim)
/// * `n` – number of oscillators
///
/// # Returns
/// Row-major (n×n) coupling matrix, non-negative, zero diagonal.
#[must_use]
pub fn decode(z: &[f64], a: &[f64], n: usize) -> Vec<f64> {
    let z_dim = z.len();
    let nn = n * n;
    let mut w = vec![0.0; nn];
    for i in 0..nn {
        let mut raw = 0.0;
        for j in 0..z_dim {
            raw += a[i * z_dim + j] * z[j];
        }
        // softplus: log(1 + exp(x))
        w[i] = softplus(raw);
    }
    // Zero diagonal
    for i in 0..n {
        w[i * n + i] = 0.0;
    }
    w
}

/// Softplus activation: log(1 + exp(x)), numerically stable.
fn softplus(x: f64) -> f64 {
    if x > 20.0 { x } else if x < -20.0 { 0.0 } else { (1.0 + x.exp()).ln() }
}

/// Finite-difference gradient of a cost function w.r.t. z.
///
/// grad_i = (cost(z + ε·e_i) - cost(z - ε·e_i)) / (2ε)
///
/// The cost is evaluated by decoding z → W and calling the cost function.
#[must_use]
pub fn finite_diff_gradient(
    z: &[f64],
    a: &[f64],
    n: usize,
    cost_fn: &dyn Fn(&[f64]) -> f64,
    epsilon: f64,
) -> Vec<f64> {
    let z_dim = z.len();
    let mut grad = vec![0.0; z_dim];
    for i in 0..z_dim {
        let mut z_plus = z.to_vec();
        z_plus[i] += epsilon;
        let mut z_minus = z.to_vec();
        z_minus[i] -= epsilon;
        let w_plus = decode(&z_plus, a, n);
        let w_minus = decode(&z_minus, a, n);
        grad[i] = (cost_fn(&w_plus) - cost_fn(&w_minus)) / (2.0 * epsilon);
    }
    grad
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_zero_z() {
        let n = 3;
        let z_dim = 4;
        let z = vec![0.0; z_dim];
        let a = vec![1.0; n * n * z_dim];
        let w = decode(&z, &a, n);
        // softplus(0) = ln(2) ≈ 0.693
        let sp0 = 2.0_f64.ln();
        for i in 0..n {
            for j in 0..n {
                if i == j {
                    assert_eq!(w[i * n + j], 0.0);
                } else {
                    assert!((w[i * n + j] - sp0).abs() < 1e-10);
                }
            }
        }
    }

    #[test]
    fn test_decode_diagonal_zero() {
        let n = 4;
        let z = vec![1.0; 3];
        let a = vec![0.5; n * n * 3];
        let w = decode(&z, &a, n);
        for i in 0..n {
            assert_eq!(w[i * n + i], 0.0);
        }
    }

    #[test]
    fn test_decode_non_negative() {
        let n = 3;
        let z = vec![-2.0, 1.0, 0.5];
        let a = vec![0.3; n * n * 3];
        let w = decode(&z, &a, n);
        for &v in &w {
            assert!(v >= 0.0, "W should be non-negative, got {v}");
        }
    }

    #[test]
    fn test_softplus_large_positive() {
        assert!((softplus(100.0) - 100.0).abs() < 1e-10);
    }

    #[test]
    fn test_softplus_large_negative() {
        assert!(softplus(-100.0) < 1e-10);
    }

    #[test]
    fn test_softplus_zero() {
        assert!((softplus(0.0) - 2.0_f64.ln()).abs() < 1e-10);
    }

    #[test]
    fn test_gradient_zero_at_minimum() {
        // Cost = sum(W), minimum at large negative z
        let n = 2;
        let z_dim = 2;
        let z = vec![-10.0; z_dim];
        let a = vec![1.0; n * n * z_dim];
        let cost = |w: &[f64]| -> f64 { w.iter().sum() };
        let grad = finite_diff_gradient(&z, &a, n, &cost, 1e-4);
        // At z = -10, softplus ≈ 0, gradient ≈ 0
        for g in &grad {
            assert!(g.abs() < 0.01, "grad={g} should be ~0 at minimum");
        }
    }

    #[test]
    fn test_gradient_direction() {
        // Cost = sum(W), gradient should be positive (increasing z increases W)
        let n = 2;
        let z_dim = 2;
        let z = vec![0.0; z_dim];
        let a = vec![1.0; n * n * z_dim];
        let cost = |w: &[f64]| -> f64 { w.iter().sum() };
        let grad = finite_diff_gradient(&z, &a, n, &cost, 1e-4);
        for g in &grad {
            assert!(*g > 0.0, "gradient should be positive, got {g}");
        }
    }
}

File: src/scpn_phase_orchestrator/ssgf/carrier.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — SSGF geometry carrier W(t)

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import TypeAlias

import numpy as np
from numpy.typing import NDArray

try:
    from spo_kernel import (
        carrier_decode_rust as _rust_decode,
    )

    _HAS_RUST = True
except ImportError:
    _HAS_RUST = False

__all__ = ["GeometryCarrier", "SSGFState"]

FloatArray: TypeAlias = NDArray[np.float64]


@dataclass
class SSGFState:
    z: FloatArray
    W: FloatArray
    cost: float
    grad_norm: float
    step: int


class GeometryCarrier:
    """SSGF outer cycle: z → W → microcycles → cost → grad → z.

    The latent vector z parameterizes a coupling geometry W(z) via a
    spectral decoder. The microcycle (Kuramoto integration) runs under
    W, produces a cost U_total, and the gradient ∂U/∂z drives z toward
    lower cost.

    This implements the autopoietic loop: geometry → dynamics → cost →
    gradient → geometry. The system produces its own coupling topology.
    """

    def __init__(
        self,
        n_oscillators: int,
        z_dim: int = 8,
        lr: float = 0.01,
        seed: int | None = None,
    ):
        self._n = n_oscillators
        self._z_dim = z_dim
        self._lr = lr
        self._rng = np.random.default_rng(seed)
        self._z: FloatArray = self._rng.normal(0, 0.1, z_dim)
        # Decoder: W = softplus(A @ z) reshaped to (n, n), zero diagonal
        # A is a fixed random projection (n*n × z_dim)
        self._A: FloatArray = self._rng.normal(
            0, 1.0 / np.sqrt(z_dim), (n_oscillators * n_oscillators, z_dim)
        )
        self._step = 0

    @property
    def z(self) -> FloatArray:
        return self._z.copy()

    @property
    def z_dim(self) -> int:
        return self._z_dim

    def decode(self, z: FloatArray | None = None) -> FloatArray:
        """Map z → coupling matrix W (n × n, non-negative, zero diagonal)."""
        if z is None:
            z = self._z
        if _HAS_RUST:
            zv = np.ascontiguousarray(z, dtype=np.float64)
            av = np.ascontiguousarray(self._A.ravel(), dtype=np.float64)
            w_flat: FloatArray = np.asarray(_rust_decode(zv, av, self._n))
            return w_flat.reshape(self._n, self._n)
        raw = self._A @ z
        # Softplus ensures non-negative coupling
        W = np.log1p(np.exp(raw)).reshape(self._n, self._n)
        np.fill_diagonal(W, 0.0)
        out: FloatArray = W
        return out

    def update(
        self,
        cost: float,
        cost_fn: Callable[[FloatArray], float] | None = None,
        epsilon: float = 1e-4,
    ) -> SSGFState:
        """One SSGF outer step: compute gradient of cost w.r.t. z, descend.

        If cost_fn is provided, uses finite differences on z.
        Otherwise records cost for external gradient computation.
        """
        self._step += 1
        grad: FloatArray = np.zeros(self._z_dim)

        if cost_fn is not None:
            for i in range(self._z_dim):
                z_plus = self._z.copy()
                z_plus[i] += epsilon
                z_minus = self._z.copy()
                z_minus[i] -= epsilon
                grad[i] = (
                    cost_fn(self.decode(z_plus)) - cost_fn(self.decode(z_minus))
                ) / (2 * epsilon)

        self._z -= self._lr * grad
        W = self.decode()
        return SSGFState(
            z=self._z.copy(),
            W=W,
            cost=cost,
            grad_norm=float(np.linalg.norm(grad)),
            step=self._step,
        )

    def reset(self, seed: int | None = None) -> None:
        rng = np.random.default_rng(seed)
        self._z = rng.normal(0, 0.1, self._z_dim)
        self._step = 0
